### PR TITLE
crush,mon: add weight-set introspection and manipulation commands

### PR DESCRIPTION
**Pull request by liewegas**
- 'osd crush tree' shows a column per weight-set (which are mapped to pools)
- json version of above shows full weight-set, including positions (plain view only shows first weight)
- mon commands to create, list, dump, remove, weight-sets
- mon command to adjust a weight in a weight-set
- many doc updates
- crush updates so adjust weight-sets sanely when items are removed from the map (by maintaining the summation property for the tree within each weight-set).

<pre>
$ ceph osd crush tree
ID WEIGHT  foo     TYPE NAME     
-1 4.00000         root default  
-2 4.00000 3.00000     host gnit 
 0 1.00000 0               osd.0 
 1 1.00000 1.00000         osd.1 
 2 1.00000 1.00000         osd.2 
 3 1.00000 1.00000         osd.3 
$ ceph osd crush tree -f json-pretty
[
    [
        {
            "id": -1,
            "name": "default",
            "type": "root",
            "type_id": 10,
            "children": [
                -2
            ]
        },
        {
            "id": -2,
            "name": "gnit",
            "type": "host",
            "type_id": 1,
            "pool_weights": {
                "foo": [
                    3.000000
                ]
            },
            "children": [
                3,
                2,
                1,
                0
            ]
        },
        {
            "id": 0,
            "name": "osd.0",
            "type": "osd",
            "type_id": 0,
            "crush_weight": 1.000000,
            "depth": 2,
            "pool_weights": {
                "foo": [
                    0.000000
                ]
            }
        },
        {
            "id": 1,
            "name": "osd.1",
            "type": "osd",
            "type_id": 0,
            "crush_weight": 1.000000,
            "depth": 2,
            "pool_weights": {
                "foo": [
                    1.000000
                ]
            }
        },
        {
            "id": 2,
            "name": "osd.2",
            "type": "osd",
            "type_id": 0,
            "crush_weight": 1.000000,
            "depth": 2,
            "pool_weights": {
                "foo": [
                    1.000000
                ]
            }
        },
        {
            "id": 3,
            "name": "osd.3",
            "type": "osd",
            "type_id": 0,
            "crush_weight": 1.000000,
            "depth": 2,
            "pool_weights": {
                "foo": [
                    1.000000
                ]
            }
        }
    ],
    []
]
</pre>

### CODE REVIEW

**Comment by liewegas** [comment]
@idryomov another semantic change here, this time with choose_args. there is now a fallback to a choose_arg index of -1 if there isn't a pool-specific choose_arg set.

This still RFC/WIP.  @jdurgin does this seem ok?

- 'ceph osd crush weight-set create-compat' creates a compat weight-set.  This acts like a fallback.  If you create a per-pool weight-set, that works for that pool.  Otherwise we try the compat/default one.  If that doesnt' exist either, then we use the normal crush weights.

**Comment by liewegas** [comment]
Not really sure how to do naming, though.  How shoudl teh compat weight-set show up in 'ls' output?  I used "(compat)" for the osd crush tree output, but that's weird to type for, say, the pool argument in 'ceph osd crush weight-set reweight (compat) osd.1 .99'.  I can do a separate reweight command (reweight-compat), but we still need to show something in 'ls' and 'dump'.  '(compat)' is a legal pool name, by the way.

**Comment by liewegas** [comment]
retest this please

**Comment by xiexingguo** [comment]
retest this please

**Comment by xiexingguo** [comment]
This is really a huge patch set. I haven't get all of these through.
But I think it is okay to merge this as long as the QA tests are happy and I'll do some follow-up validations later.

File: src/crush/CrushWrapper.cc [hunk 1]
```diff
@@ -2786,3 +3000,86 @@ int CrushWrapper::try_remap_rule(
...
+      for (unsigned j = 0; j < weight.size(); ++j) {
+	carg->weight_set[j].weights[i] = weight[j];
+      }
+      ldout(cct, 5) << __func__ << "  set " << id << " to " << weight
+		    << " in bucket " << b->id << dendl;
+      touched = true;
+      changed++;
+    }
+  }
+  if (touched) {
```
Comment (xiexingguo) [inline src/crush/CrushWrapper.cc]: if (changed) will be enough(hence we can drop `touched `here)

File: src/mon/OSDMonitor.cc [hunk 2]
```diff
@@ -7305,7 +7340,121 @@ bool OSDMonitor::prepare_command_impl(MonOpRequestRef op,
...
+    pending_inc.crush.clear();
+    newcrush.encode(pending_inc.crush, mon->get_quorum_con_features());
+    goto update;
+
+  } else if (prefix == "osd crush weight-set rm" ||
+	     prefix == "osd crush weight-set rm-compat") {
+    CrushWrapper newcrush;
+    _get_pending_crush(newcrush);
+    int64_t pool;
+    if (prefix == "osd crush weight-set create") {
```
Comment (xiexingguo) [inline src/mon/OSDMonitor.cc]: should be "osd crush weight-set **rm**"

**Comment by xiexingguo** [comment]
BTW, the documentation really helps!

**Comment by liewegas** [comment]
fixed those 2 items and updated the test so that it should catch the 'rm' bug.

**Comment by liewegas** [comment]
/a/sage-2017-07-21_03:24:52-rados-wip-sage-testing-distro-basic-smithi/1426654
/a/sage-2017-07-21_03:24:52-rados-wip-sage-testing-distro-basic-smithi/1426454

File: src/crush/CrushWrapper.cc [hunk 3]
```diff
@@ -1990,9 +1990,12 @@ void CrushWrapper::decode(bufferlist::iterator& blp)
 	      ::decode(weight_set->weights[l], blp);
 	  }
 	  ::decode(arg->ids_size, blp);
-	  arg->ids = (__s32 *)calloc(arg->ids_size, sizeof(__s32));
```
Comment (idryomov) [inline src/crush/CrushWrapper.cc]: Should we do the same for weight_set calloc (slightly above) and get rid of arg->weight_set_size == 0 check in get_choose_arg_weights()?  I wondered why it's there...

The kernel client checks both: https://github.com/ceph/ceph-client/blob/33e9c8dbfbcef8e4cda8e43a445e692ab7e0d8c0/net/ceph/osdmap.c#L222

Comment (liewegas) [inline src/crush/CrushWrapper.cc]: yeah, fixing!

File: src/crush/CrushWrapper.cc [hunk 4]
```diff
@@ -2136,15 +2136,17 @@ void CrushWrapper::decode(bufferlist::iterator& blp)
...
-	  arg->weight_set = (crush_weight_set*)calloc(
-	    arg->weight_set_size, sizeof(crush_weight_set));
-	  for (__u32 k = 0; k < arg->weight_set_size; k++) {
-	    crush_weight_set *weight_set = &arg->weight_set[k];
-	    ::decode(weight_set->size, blp);
-	    weight_set->weights = (__u32*)calloc(
-	      weight_set->size, sizeof(__u32));
-	    for (__u32 l = 0; l < weight_set->size; l++)
-	      ::decode(weight_set->weights[l], blp);
+	  if (arg->weight_set_size) {
```
Comment (idryomov) [inline src/crush/CrushWrapper.cc]: What about arg->weight_set_size == 0 check in get_choose_arg_weights()?  get_choose_arg_ids() doesn't have a corresponding arg->ids_size == 0 check, so we shouldn't need a weight_set one now either.

Comment (liewegas) [inline src/crush/CrushWrapper.cc]: Ah, yep.. that's now true.  fixed!

**Comment by liewegas** [comment]
2017-07-23T02:37:55.096 INFO:tasks.ceph.osd.2.smithi065.stderr:/build/ceph-12.1.1-380-g5e8fa3e/src/osd/PrimaryLogPG.cc: 1845: FAILED assert(!cct->_conf->osd_debug_misdirected_ops)
2017-07-23T02:37:55.097 INFO:tasks.ceph.osd.2.smithi065.stderr:2017-07-23 02:37:55.090590 7f76325b2700 -1 osd.2 pg_epoch: 83 pg[3.7( v 83'1 (0'0,83'1] local-lis/les=80/83 n=1 ec=73/73 lis/c 80/80 les/c/f 83/83/0 73/80/73) [2,1] r=0 lpr=82 luod=0'0 lua=0'0 crt=83'1 lcod 0'0 mlcod 0'0 active+clean] do_op 3.7 does not contain 3:ee71ac71:::benchmark_data_smithi065_17586_object26:head pg_num 26 hash 8e358e77

/a/sage-2017-07-23_02:11:52-rados-wip-weight-set-distro-basic-smithi/1433151

**Comment by liewegas** [comment]
ok that failure was unrelated, http://tracker.ceph.com/issues/20754

**Comment by liewegas** [comment]
@idryomov updated patch look ok?  If so, I think this is ready to merge!

**Comment by xiexingguo** [comment]
👍

**Comment by idryomov** [comment]
The parts I looked at look good.

**Comment by xiexingguo** [comment]
> The parts I looked at look good.

+1 @liewegas Hope this can merge ASAP so I can continue with https://github.com/ceph/ceph/pull/16388

**Pull request merged**